llvm/lib/Target/AArch64/AArch64ISelLowering.cpp

```diff
@@ -283,6 +283,24 @@ static bool isZeroingInactiveLanes(SDValue Op) {
     switch (Op.getConstantOperandVal(0)) {
     default:
       return false;
+
+    case Intrinsic::aarch64_sve_uzp1:
+      return isZeroingInactiveLanes(Op.getOperand(1)) &&
+             isZeroingInactiveLanes(Op.getOperand(2));
```

Comment:
https://reviews.llvm.org/D141469#4057420 has some comments by @paulwalker-arm about how the actual "block" of type d/s/h/b is copied unmodified and there may be some bits in there that don't belong to the predicate type.

Reply:
Oh, I see, yes, you're right.

Reply:
I'm not so sure this is the correct place for such code because the function related to zeroing through construction rather than propagation. It is a bit of a kludge but I find comfort in knowing it only handles a specific set of intrinsics that we know have no special handling.  Once you start recursing down the DAG I start to feel less comfortable.

That said, `aarch64_sve_uzp1` is also a weird intrinsic because for predicates only the `nxv16i1` case is tied by design to a specific instruction (i.e. it's not one of the intrinsics where, at least as of today, we've attached any meaning to the invisible lanes).  I only maintained the support for other predicate types when creating https://reviews.llvm.org/D142065 because it seemed useful to maintain it as a way to logically interleave two vectors.  But we have dedicated target neutral intrinsics for that now.

So I guess my question is in what circumstance do you need to support `to_svbool(sve.uzp1(val))`? I wasn't expecting to see this from ACLE code for example.  I'm not against the patch but I would like to understand the rational more.

Reply:
We have some code similar to minimal C example below, implemented using intrinsics. The IR example would be `chainedUzp1_to_svbool` in the tests. The result of icmp is reinterpreted and combined using uzp1 and to/from svbool until we get `vscale x 16 x i1`.

If this function is not an appropriate place for this code then would it make sense to do it in instcombine by simplifying this kind of pattern to a concat using the llvm.vector.insert intrinsic?

```
void test(uint8_t *x, uint8_t *y, uint32_t v, int n) {
  for (int i = 0; i < n; i++) {
    x[i] = y[i]==v;
  }
}
```

Reply:
Posted an instcombine version: https://github.com/llvm/llvm-project/pull/81069

Reply:
I see, thanks for the info.  I certainly prefer the idea of having this as an instcombine so I'll take a look at the other PR.  Looking at the IR I do wonder though if you're just emitting the wrong intrinsics and rather than using `sve.uzp1` for everything you could use the larger element variants (e.g. `sve.uzp1.b16`) that operate on `vscale x 16 x i1` and thus you remove the need for `sve.convert.from.svbool` intrinsics.